zcash_history/src/node_data.rs

```diff
@@ -180,27 +155,64 @@ impl NodeData {
 
     /// Convert to byte representation.
     pub fn to_bytes(&self) -> Vec<u8> {
-        let mut buf = [0u8; MAX_NODE_DATA_SIZE];
-        let pos = {
-            let mut cursor = std::io::Cursor::new(&mut buf[..]);
-            self.write(&mut cursor).expect("Cursor cannot fail");
-            cursor.position() as usize
-        };
-
-        buf[0..pos].to_vec()
+        crate::V1::to_bytes(self)
     }
 
     /// Convert from byte representation.
     pub fn from_bytes<T: AsRef<[u8]>>(consensus_branch_id: u32, buf: T) -> std::io::Result<Self> {
-        let mut cursor = std::io::Cursor::new(buf);
-        Self::read(consensus_branch_id, &mut cursor)
+        crate::V1::from_bytes(consensus_branch_id, buf)
     }
 
     /// Hash node metadata
     pub fn hash(&self) -> [u8; 32] {
-        let bytes = self.to_bytes();
+        crate::V1::hash(self)
+    }
+}
+
+/// V2 node metadata.
+#[derive(Debug, Clone, Default)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct V2 {
+    /// The V1 node data retained in V2.
+    pub v1: NodeData,
```

Comment:
Clever

Reply:
This also looks good for Zebra.

We can take our `block_to_history_node` function that produces a `zcash_history::NodeData`, and wrap it in a `block_to_history_node_v2` function that returns `zcash_history::V2`:
https://github.com/ZcashFoundation/zebra/pull/2227/files#diff-ccc89b2ef027c45e027753030442d684005ebf402fcc69be39c70ee6ed193773R230

We might want to rename the legacy function to `_v1`, but we can do that in our Orchard-specific PR.

Reply:
As an aside, adding new fields is a lot simpler than deleting fields.

We deleted the per-spend anchor for sapling using marker traits, and it got slightly complicated:

https://github.com/ZcashFoundation/zebra/blob/4aecf0360724e50d51f31ab15e86277e5b8bbd32/zebra-chain/src/sapling/shielded_data.rs#L267-L291

https://github.com/ZcashFoundation/zebra/blob/4aecf0360724e50d51f31ab15e86277e5b8bbd32/zebra-chain/src/sapling/spend.rs#L76-L89

But having a shared implementation is still really nice.

Reply:
Yeah, there will likely be some duplication if we ever change or delete a field (e.g. when Sapling is eventually tuned off). But that's unlikely to happen for quite some time.